File: backend/internal/chat/adapters/httpapi/stream.go
package httpapi

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"net/http"

	"github.com/google/uuid"

	"github.com/corsi/backend/internal/chat/app"
	"github.com/corsi/backend/internal/chat/domain"
	"github.com/corsi/backend/internal/platform/apierror"
)

type sendMessageRequest struct {
	Content string `json:"content"`
	// References is the capability selection the composer attached to this
	// turn. Additive: a client that never sends it gets the behaviour it has
	// always had, and an explicit `[]` means the same as omitting it.
	//
	// See referenceRequest for why the wire shape carries no label.
	References []referenceRequest `json:"references"`
	// ContextReferences are the entities this turn is about. Additive in
	// exactly the same way: a client that never sends it behaves as it
	// always did.
	ContextReferences []contextReferenceRequest `json:"context_references"`
}

// referenceRequest is one attached capability, as the client may state it.
//
// Identity only. The display label of the permanent record is written by
// the server from its own registry — a client that could name a capability
// must not also get to name it whatever it likes in the transcript. The
// decoder rejects unknown fields, so sending a label is a 400 rather than a
// value that is silently thrown away, which is the honest answer to a
// client asserting something the server will not honour.
type referenceRequest struct {
	Kind string `json:"kind"`
	ID   string `json:"id"`
}

// contextReferenceRequest is one attached ENTITY, as the client may state
// it.
//
// ── Why it carries no label either, and why that matters more here ─────
// Same rule as referenceRequest, with a sharper edge. A tool label the
// client could author would be a lie in a transcript; an ENTITY label the
// client could author would also be a lie in the MODEL'S CONTEXT — the
// block that tells the agent what it is looking at. So the label is
// resolved server-side from the owning module and whatever the client sent
// is not read. The decoder rejects unknown fields, so sending one is a 400
// rather than a value silently discarded.
type contextReferenceRequest struct {
	Type string `json:"type"`
	ID   string `json:"id"`
}

// toContextReferences converts the wire shape, carrying identity only.
// Resolution, workspace scoping and the label all happen in the
// application layer — see app.admitContextReferences.
func toContextReferences(in []contextReferenceRequest) []domain.ContextReference {
	if len(in) == 0 {
		return nil
	}
	out := make([]domain.ContextReference, 0, len(in))
	for _, r := range in {
		out = append(out, domain.ContextReference{
			Type: domain.ContextReferenceType(r.Type),
			ID:   r.ID,
		})
	}
	return out
}

// sendMessage streams the model's reply as Server-Sent Events.
//
// The stream is opened lazily — on the first byte the model produces, not
// when the handler starts. That keeps the ordinary failures (unknown
// conversation, unreadable credential, provider refusing the key) as real
// HTTP status codes with the same error body as every other endpoint,
// because none of them can happen after a token has been sent. Once the
// stream is open there is no status code left to give, so failures are
// reported as an `error` frame inside it.
//
// Frames:
//
//	event: reasoning  data: {"text":"…"}
//	event: delta      data: {"text":"…"}
//	event: tool       data: {"call_id":"…","name":"…","status":"running|ok|error",…}
//	event: done       data: {"message":{…}}
//	event: error      data: {"error":{"code":"…","message":"…"}}
//
// `reasoning` frames, when a reasoning model produces them, all arrive
// before the first `delta`.
//
// `tool` frames arrive only on a turn that runs a tool, in pairs: one
// `running` when a call starts, one `ok` or `error` when it ends. A turn
// with no tools emits none, so the four frames that existed before are
// exactly the four such a turn still sends — and a client that has never
// heard of `tool` ignores the event name, which is how SSE is specified to
// behave.
func (h *Handler) sendMessage(w http.ResponseWriter, r *http.Request) {
	ws, ok := h.workspaceID(w, r)
	if !ok {
		return
	}
	id, ok := h.pathID(w, r)
	if !ok {
		return
	}
	var req sendMessageRequest
	if !h.decode(w, r, &req) {
		return
	}

	sink := newSSESink(w, r.Context(), h.log)
	// The receipt is read after the turn has been written back, using a
	// context that outlives the request the way the write-back does: a
	// client that hung up still gets its audit row, and the frame is simply
	// not delivered.
	sink.receipts = func(messageID uuid.UUID) (domain.WriteReceipt, error) {
		got, err := h.svc.WriteReceipts(context.WithoutCancel(r.Context()), ws, []uuid.UUID{messageID})
		if err != nil {
			return domain.WriteReceipt{}, err
		}
		return got[messageID], nil
	}
	// The read-side twin, on the same terms.
	sink.readReceipts = func(messageID uuid.UUID) (domain.ReadReceipt, error) {
		got, err := h.svc.ReadReceipts(context.WithoutCancel(r.Context()), ws, id, []uuid.UUID{messageID})
		if err != nil {
			return domain.ReadReceipt{}, err
		}
		return got[messageID], nil
	}

	msg, err := h.svc.SendMessage(r.Context(), app.SendMessageInput{
		WorkspaceID:       ws,
		ConversationID:    id,
		Content:           req.Content,
		References:        toTurnReferences(req.References),
		ContextReferences: toContextReferences(req.ContextReferences),
	}, sink)

	if err != nil {
		if !sink.opened {
			h.writeDomainErr(w, err)
			return
		}
		sink.writeError(err)
		return
	}
	sink.writeDone(msg)
}

// toTurnReferences lifts the wire shape into the domain's, unchanged and
// unvalidated.
//
// Deliberately dumb: whether a kind exists, whether the tool is real and
// whether this agent may use it are all questions for the application
// layer, which is the only place that can see the registry and the grants.
// A transport that answered any of them would be a second authorization
// point, free to disagree with the first.
//
// An empty list becomes a nil slice, because omitted and empty are the same
// input — see app.SendMessageInput.References.
func toTurnReferences(in []referenceRequest) []domain.TurnReference {
	if len(in) == 0 {
		return nil
	}
	out := make([]domain.TurnReference, 0, len(in))
	for _, r := range in {
		out = append(out, domain.TurnReference{
			Kind: domain.ReferenceKind(r.Kind),
			ID:   r.ID,
		})
	}
	return out
}

// --- SSE sink -------------------------------------------------------------

type sseSink struct {
	w      http.ResponseWriter
	rc     *http.ResponseController
	ctx    context.Context
	log    *slog.Logger
	opened bool
	// receipts resolves the finished turn's write receipt. A function
	// rather than a service handle so the sink keeps knowing nothing about
	// the application layer, which is the arrangement every other frame
	// already has.
	receipts     func(messageID uuid.UUID) (domain.WriteReceipt, error)
	readReceipts func(messageID uuid.UUID) (domain.ReadReceipt, error)
}

func newSSESink(w http.ResponseWriter, ctx context.Context, log *slog.Logger) *sseSink {
	// ResponseController rather than a direct http.Flusher assertion: the
	// router wraps the writer twice (request logging, then metrics), and
	// the controller unwraps through both to reach the real one.
	return &sseSink{w: w, rc: http.NewResponseController(w), ctx: ctx, log: log}
}

func (s *sseSink) open() {
	if s.opened {
		return
	}
	h := s.w.Header()
	h.Set("Content-Type", "text/event-stream")
	h.Set("Cache-Control", "no-cache")
	h.Set("Connection", "keep-alive")
	// Without this, an nginx in front of the API buffers the whole response
	// and the reply lands in one lump at the end — the stream still works,
	// but nothing about it looks streamed.
	h.Set("X-Accel-Buffering", "no")
	s.w.WriteHeader(http.StatusOK)
	s.opened = true
	_ = s.rc.Flush()
}

// event writes one SSE frame. Payloads are JSON-encoded, which also solves
// the framing problem: SSE terminates a frame on a blank line, and JSON
// escaping guarantees the encoded data contains no raw newline.
func (s *sseSink) event(name string, payload any) error {
	if err := s.ctx.Err(); err != nil {
		return err // client already gone; don't bother writing
	}
	s.open()

	data, err := json.Marshal(payload)
	if err != nil {
		return fmt.Errorf("marshal %s frame: %w", name, err)
	}
	if _, err := fmt.Fprintf(s.w, "event: %s\ndata: %s\n\n", name, data); err != nil {
		return err
	}
	// Flush is where a dead connection actually surfaces — Write alone
	// buffers and can succeed long after the reader is gone.
	return s.rc.Flush()
}

// Reasoning, Delta and Tool implement app.TurnSink.
func (s *sseSink) Reasoning(text string) error {
	return s.event("reasoning", map[string]string{"text": text})
}

func (s *sseSink) Delta(text string) error {
	return s.event("delta", map[string]string{"text": text})
}

// Tool forwards a tool call's progress. The payload is the app-layer event
// verbatim: it already carries only the names and outcomes a transcript
// shows, and no arguments or results.
func (s *sseSink) Tool(ev app.ToolEvent) error {
	return s.event("tool", ev)
}

func (s *sseSink) writeDone(msg *domain.Message) {
	// msg is nil only when the post-turn write-back failed; the reply was
	// still delivered, so the stream closes normally and the client keeps
	// what it rendered.
	payload := map[string]any{"message": msg}
	// ── The turn's own receipt, in the frame that ends it ───────────
	//
	// So a live client never has to infer from the prose it just rendered
	// whether anything changed. A turn with no write executed carries
	// `write_receipt.executed == 0`, which is the statement that would have
	// contradicted "8 transações importadas" when a model said it after
	// calling nothing.
	//
	// Built from the same audit rows the transcript reads, so the live view
	// and a reload cannot disagree.
	if msg != nil && s.receipts != nil {
		if r, err := s.receipts(msg.ID); err == nil {
			payload["write_receipt"] = r
		} else {
			s.log.Warn("write receipt for done frame", "message_id", msg.ID, "err", err)
		}
	}
	// ── The read-side twin, and the reason it is not optional ──────
	//
	// A turn that read nothing external carries
	// `read_receipt.status == "NO_EXTERNAL_READ"`, which is the statement
	// that would have contradicted "1.535 seguidores, 40.055 views" when a
	// model produced those figures after calling nothing.
	//
	// Emitted for every turn, including the ones that read nothing:
	// absence has to be a value, or a surface renders nothing at all for
	// exactly the turn that invented a number.
	if msg != nil && s.readReceipts != nil {
		if r, err := s.readReceipts(msg.ID); err == nil {
			payload["read_receipt"] = r
		} else {
			s.log.Warn("read receipt for done frame", "message_id", msg.ID, "err", err)
		}
	}
	if err := s.event("done", payload); err != nil {
		s.log.Debug("write done frame", "err", err)
	}
}

func (s *sseSink) writeError(err error) {
	apiErr := domainAPIError(err)
	if apiErr == nil {
		s.log.Error("unhandled error mid-stream", "err", err)
		apiErr = apierror.ErrInternal
	}
	if werr := s.event("error", map[string]any{"error": apiErr}); werr != nil {
		s.log.Debug("write error frame", "err", werr)
	}
}

File: frontend/src/modules/agents/api/stream.ts
/**
 * Streaming reader for POST /chat/conversations/{id}/messages.
 *
 * ── Why not EventSource ────────────────────────────────────────────────
 * The browser's `EventSource` can only issue a GET and cannot set request
 * headers, so it can carry neither the message body nor `X-Workspace-Id`.
 * Every request in this app is workspace-scoped, so the built-in client is
 * unusable here. `fetch` + `ReadableStream` gives us POST, headers, and
 * cancellation — at the cost of parsing the event framing ourselves, which
 * is the small block at the bottom of this file.
 *
 * ── The backend's two error modes ──────────────────────────────────────
 * The stream opens lazily, on the model's first token. Anything that fails
 * before then (unknown conversation, missing key, provider refusing the
 * request) arrives as an ordinary JSON error with a real status code, and
 * is thrown as `ApiError` exactly like any other endpoint. Once tokens are
 * flowing there is no status left to send, so a later failure arrives as an
 * `error` frame inside the stream and surfaces through `onError`.
 */

import { API_BASE, ApiError, type ApiErrorBody } from "@/lib/api/client";
import { getApiWorkspaceId } from "@/lib/api/workspace";
import type { ApiMessage } from "./conversations";

/**
 * One tool call's progress, live.
 *
 * Two frames per call: `running` when it starts, then the outcome —
 * `ok`, `error`, or `not_executed` for a call the runtime refused before
 * it ran. The same vocabulary the audit row uses, so the live view and a
 * reload cannot describe one event differently.
 * Correlate them by `call_id` — a round may run several tools at once, and
 * their frames interleave.
 *
 * No arguments and no result: the live channel says what is happening, and
 * the payloads are behind the audit read the user asks for.
 */
export interface ToolFrame {
  call_id: string;
  name: string;
  status: "running" | "ok" | "error" | "not_executed";
  error_code?: string;
  duration_ms?: number;
}

/**
 * What the SYSTEM knows a turn did, as opposed to what the model said.
 *
 * A live financial agent once answered "8 transações importadas" in a turn
 * that called nothing, against a ledger holding zero transactions. The
 * runtime was right and the person was misinformed, because the absence of
 * a tool call is invisible in prose.
 *
 * So the answer is explicit for every assistant turn, including the turns
 * where nothing ran: `executed === 0` is a statement, not a missing field.
 * Anything that tells a user a change happened must read this, never the
 * text beside it.
 */
export interface WriteExecution {
  tool_call_id: string;
  capability: string;
  status: "REQUESTED" | "EXECUTED" | "FAILED" | "NOT_EXECUTED";
  occurred_at: string;
  duration_ms: number;
  error_code?: string;
}

export interface WriteReceipt {
  message_id: string;
  writes: WriteExecution[];
  executed: number;
  failed: number;
  refused: number;
}

/** One call to a system this product does not own. Metadata only — never
 *  what the call returned. See domain/read_receipt.go. */
export interface ExternalRead {
  tool_call_id: string;
  capability: string;
  /** The capability's owner, derived from its name: "meta_threads". */
  source: string;
  status: "VERIFIED_EXTERNAL_READ" | "FAILED_EXTERNAL_READ" | "NO_EXTERNAL_READ";
  occurred_at: string;
  duration_ms: number;
  error_code?: string;
}

/**
 * Whether this turn read anything outside the product.
 *
 * The read-side twin of WriteReceipt, and the answer to a turn that
 * reported a follower count having called nothing. Derived from execution
 * records; it never depends on what the model wrote.
 */
export interface ReadReceipt {
  message_id: string;
  reads: ExternalRead[];
  status: "VERIFIED_EXTERNAL_READ" | "FAILED_EXTERNAL_READ" | "NO_EXTERNAL_READ";
  verified: number;
  failed: number;
  /** Whether the agent could have read externally at all. A presentation
   *  gate, not a claim about the turn. */
  available: boolean;
}

export interface StreamHandlers {
  /**
   * One increment of the model's chain of thought. Only reasoning models
   * emit these, and they all arrive before the first `onDelta`.
   */
  onReasoning: (text: string) => void;
  /** One increment of answer text. Append it; do not replace. */
  onDelta: (text: string) => void;
  /**
   * A tool call starting or finishing. Optional, and absent handlers are
   * simply not called — a turn with no tools never emits one.
   */
  onTool?: (frame: ToolFrame) => void;
  /**
   * Clean finish. Carries the persisted message and the turn's write
   * receipt, so a live client never has to read the prose to find out
   * whether anything changed.
   */
  onDone: (
    message: ApiMessage | null,
    receipt?: WriteReceipt,
    readReceipt?: ReadReceipt,
  ) => void;
  /** Failure after the stream opened. The turn is over. */
  onError: (error: ApiError) => void;
}

/**
 * Sends a message and streams the reply.
 *
 * Resolves when the stream ends, for any reason. Throws only for failures
 * that happened before the stream opened — those still have a status code
 * and belong to the caller's normal error handling. Aborting via `signal`
 * resolves quietly: the backend persists whatever text it had already sent.
 */
/**
 * One capability attached to the turn being sent.
 *
 * Identity only. The label of the stored record is written by the server
 * from its own registry — sending one is rejected outright rather than
 * ignored, because a client that could name a capability must not also get
 * to author how the transcript will describe it forever after.
 */
/**
 * One thing the user attached to a turn, as the wire states it.
 *
 * Identity only, and never a label: the words that reach the permanent
 * record are written by the server from its own registry — see
 * chat/domain/reference.go. The decoder rejects unknown fields, so sending
 * a label is a 400 rather than a value silently discarded.
 *
 * `integration` names a PROVIDER by the namespace its capabilities are
 * named under (`github`). The server expands it against this agent's
 * grants and freezes the individual capabilities that were really exposed,
 * which is what keeps an old turn from appearing to have reached a
 * capability that shipped after it ran.
 */
export interface SendReference {
  kind: "integration";
  id: string;
}

export async function streamMessage(
  conversationId: string,
  content: string,
  handlers: StreamHandlers,
  signal?: AbortSignal,
  references?: readonly SendReference[],
): Promise<void> {
  return streamTurn(
    `/chat/conversations/${conversationId}/messages`,
    // The field is omitted entirely when nothing was attached. An explicit
    // `[]` means the same thing to the server, but a body that carries only
    // what it has to is the one a pre-Batch-4 backend would also accept.
    references && references.length > 0 ? { content, references } : { content },
    handlers,
    signal,
  );
}

/**
 * Continue a turn that stopped with work already done.
 *
 * ── Why this is not `streamMessage("Try again")` ───────────────────────
 * Because that is what caused the first User Beta incident. A turn created
 * a Room and a list, hit the tool-round ceiling before adding the items,
 * and the only route the product offered was typing into the composer.
 * "Try again" is a new question, and "again" means start over: the next
 * turn created a SECOND Room and a SECOND list.
 *
 * A resume adds no question. It names the interrupted turn, and the server
 * answers the ORIGINAL one again — this time knowing which entities the
 * first attempt already created. See chat/app/resume.go.
 */
export async function streamResume(
  conversationId: string,
  messageId: string,
  handlers: StreamHandlers,
  signal?: AbortSignal,
): Promise<void> {
  return streamTurn(
    `/chat/conversations/${conversationId}/resume`,
    { message_id: messageId },
    handlers,
    signal,
  );
}

/**
 * The transport both turn-producing routes share.
 *
 * One function because the SSE handling below is subtle — partial frames,
 * multi-byte boundaries, abort-versus-failure — and a second copy of it
 * would be where a resume quietly stopped delivering `done`, which is the
 * frame carrying the receipts.
 */
async function streamTurn(
  path: string,
  body: unknown,
  handlers: StreamHandlers,
  signal?: AbortSignal,
): Promise<void> {
  const res = await fetch(`${API_BASE}${path}`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "X-Workspace-Id": getApiWorkspaceId(),
      Accept: "text/event-stream",
    },
    body: JSON.stringify(body),
    signal,
  });

  if (!res.ok) {
    // Pre-stream failure: a normal JSON error body with a status code.
    const text = await res.text();
    let body: ApiErrorBody | null = null;
    try {
      const parsed: unknown = JSON.parse(text);
      if (parsed && typeof parsed === "object" && "error" in parsed) {
        body = parsed as ApiErrorBody;
      }
    } catch {
      /* non-JSON error body — fall through to the status text */
    }
    throw new ApiError(res.status, body, res.statusText);
  }

  if (!res.body) {
    throw new ApiError(0, null, "this browser cannot read streamed responses");
  }

  const reader = res.body.getReader();
  // `stream: true` is what makes multi-byte characters safe: a chunk
  // boundary can land in the middle of an "ã", and a stateless decoder
  // would emit a replacement character for each half.
  const decoder = new TextDecoder("utf-8");
  let buffer = "";

  try {
    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });

      // Frames are separated by a blank line. The trailing element is
      // whatever arrived after the last separator — an incomplete frame
      // that stays buffered until the rest of it shows up.
      const frames = buffer.split("\n\n");
      buffer = frames.pop() ?? "";

      for (const frame of frames) {
        dispatch(frame, handlers);
      }
    }
    // A well-formed stream ends on a frame boundary, so the remainder is
    // normally empty. Flush it anyway rather than dropping a final event.
    if (buffer.trim()) dispatch(buffer, handlers);
  } catch (err) {
    // An abort is the user pressing stop, not a failure to report.
    if (signal?.aborted) return;
    throw err;
  } finally {
    reader.releaseLock();
  }
}

/* ── frame parsing ───────────────────────────────────────────────────── */

interface DeltaPayload {
  text: string;
}
interface DonePayload {
  message: ApiMessage | null;
  write_receipt?: WriteReceipt;
  read_receipt?: ReadReceipt;
}
type ErrorPayload = ApiErrorBody;

/** Parses one SSE frame and routes it to the matching handler. */
function dispatch(frame: string, handlers: StreamHandlers): void {
  let event = "message";
  const dataLines: string[] = [];

  for (const line of frame.split("\n")) {
    if (line.startsWith(":")) continue; // keep-alive comment
    if (line.startsWith("event:")) {
      event = line.slice(6).trim();
    } else if (line.startsWith("data:")) {
      dataLines.push(line.slice(5).trim());
    }
  }
  if (dataLines.length === 0) return;

  let payload: unknown;
  try {
    payload = JSON.parse(dataLines.join("\n"));
  } catch {
    return; // a frame we can't read shouldn't kill a working stream
  }

  switch (event) {
    case "reasoning":
      handlers.onReasoning((payload as DeltaPayload).text ?? "");
      break;
    case "delta":
      handlers.onDelta((payload as DeltaPayload).text ?? "");
      break;
    case "tool":
      handlers.onTool?.(payload as ToolFrame);
      break;
    case "done": {
      const done = payload as DonePayload;
      handlers.onDone(done.message ?? null, done.write_receipt, done.read_receipt);
      break;
    }
    case "error": {
      const body = payload as ErrorPayload;
      handlers.onError(new ApiError(502, body, "the stream failed"));
      break;
    }
  }
}
